src/openvslam/data/map_database.cc

```diff
@@ -71,6 +71,32 @@ std::vector<keyframe*> map_database::get_all_keyframes() const {
     return keyframes;
 }
 
+std::vector<keyframe*> map_database::get_close_keyframes(const Mat44_t& pose,
+                                                         const double distance_threshold,
+                                                         const double cos_angle_threshold) const {
```

Comment:
Why the angular threshold? I'm concerned that then you could not relocalize in a space where you're driving the robot in the opposite direction that the SLAM model was captured from. 

If this is a large contributor, would it then make sense to find the closest keyframes not by distance/angle, but rather by overlapping regions of the features contained at that keyframe?

Reply:
> Why the angular threshold? I'm concerned that then you could not relocalize in a space where you're driving the robot in the opposite direction that the SLAM model was captured from.

If the camera were facing the opposite direction, the angle would be 180 degrees. 

> If this is a large contributor, would it then make sense to find the closest keyframes not by distance/angle, but rather by overlapping regions of the features contained at that keyframe?

Are you going to calculate it for all keyframes? I think this is an overthought first step. Without any additional implementation, candidates that do not meet the minimum thresholds will be eliminated immediately.
https://github.com/OpenVSLAM-Community/openvslam/blob/2e2c4eaa96533317d759c3a9198bc0a5fc659e90/src/openvslam/module/relocalizer.cc#L54

Reply:
These thresholds are actually represents the maximum distance between the requested pose and selected keyframes when doing nearest keyframe search. 

> I'm concerned that then you could not relocalize in a space where you're driving the robot in the opposite direction that the SLAM model was captured from.

That is why the angular threshold was introduced - to not include the unnecessary keyframes into candidates list. If the map was recorded by camera looking only on north, but you are trying to relocalize on south, it will fail in any case.

We previously have agreed that we are weeding out the candidates having minimum required distance and angle (closest candidates). If we will try to search keyframes in another way by nearest feature scope in it, it looks like we will do the same as `bow_db_->acquire_relocalization_candidates()` does and thus fall into a "usual relocalization" where the input pose is not being taken into account. So, that is why all these steps with searching of nearest candidates were introduced.

Reply:
> it will fail in any case
Will it? If you're pointing south but frames were captured moving north, wouldn't some of the further back north-facing keyframes contain features visible on the other side by the now south facing camera? 

If we make relocalization tied to the direction of mapping, we're in some deep trouble for reliability unless we heavily document the fact that its expected you drive up and down each area from all possible directions. Maybe not something to deal with in this PR, but that's concerning for reliability.

Reply:
Created https://github.com/OpenVSLAM-Community/openvslam/issues/108.